[DomCrawler] Abstract URI logic and crawl images

valeriangalliat opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #12429 |
| License | MIT |
| Doc PR | symfony/symfony-docs#4971 |

This is a backward-compatible version of #13620, and a rebase of #13649 on current `master`.

valeriangalliat committed:
[DomCrawler] Abstract URI-related logic

All the URI parsing logic is externalized in the AbstractUriElement
class. This class have two abstract methods:

* setNode: validate the DOMElement node according to the concrete class
  rules, and set $this->node.
* getRawUri: get the raw URI from $this->node.

The Link classs now extends AbstractUriElement.

This refactor is desirable for #12429.

valeriangalliat committed:
[DomCrawler] Added ability to crawl images

A new Image class is added, extending AbstractUriElement, to leverage
URI methods for the HTML img src attribute.

Two methods are added to the Crawler class, image and images, that are
the equivalent of link and links for images.

fabpot commented on src/Symfony/Component/DomCrawler/CHANGELOG.md:
```diff
@@ -1,6 +1,12 @@
 CHANGELOG
 =========
 
+2.7.0
```
should be 3.1.0 :)

valeriangalliat commented:
About the tests, the only error I can see is about the framework bundle, any idea what I can do?

fabpot commented:
:+1:

jakzal commented on src/Symfony/Component/DomCrawler/Image.php:
```diff
@@ -0,0 +1,37 @@
+{
+    public function __construct(\DOMElement $node, $currentUri)
+    {
+        parent::__construct($node, $currentUri);
```
Why is this constructor overriden?

stof replied:
@jakzal the goal is to forbid passing the third argument. An Image is always a GET request

valeriangalliat replied:
@jakzal, @stof, I can explicitly pass `GET` as third argument to make it more obvious maybe?

jakzal replied:
@valeriangalliat yes, I think it would be more clear. I'm not gonna block this PR if you don't do this though ;)

jakzal commented on src/Symfony/Component/DomCrawler/AbstractUriElement.php:
```diff
@@ -0,0 +1,214 @@
+    protected $currentUri;
+
+    /**
+     * Constructor.
```
This comment can be removed.

stloyd replied:
Useless comment IMO.

valeriangalliat replied:
Same opinion, but kept consistency with https://github.com/symfony/symfony/blob/8573385a83b404bb16e361e93edc2b67b054edbf/src/Symfony/Component/DomCrawler/Link.php#L37

fabpot replied:
Can be removed safely :)

valeriangalliat replied:
@fabpot sure, I see it's present also in `Crawler.php`, `Form.php` and `Field/FormField.php`, should I remove it there too? Or I just edit `AbstractUriElement.php` and `Crawler.php` that are the only ones I changed in this PR?

stof replied:
We could remove them in the existing class, but not in this PR. It should be done in older branches too.

fabpot replied:
Just remove them in your own code for now.

valeriangalliat replied:
Sure, done!

jakzal commented on src/Symfony/Component/DomCrawler/UriElementInterface.php:
```diff
@@ -0,0 +1,39 @@
+    public function getNode();
+
+    /**
+     * Gets the method associated with this element.
```
I'd mentioned it's an "http method" not just "method".

valeriangalliat commented:
@jakzal I got two other comments notification that you seem to have removed, I assume everything's okay? It was about existing code that I just moved around but still, if we can improve stuff here, I can update the PR. :)

jakzal commented:
@valeriangalliat yeah, sorry for the noise. I noticed the code was moved and removed those comments.

jakzal commented on src/Symfony/Component/DomCrawler/Crawler.php:
```diff
@@ -731,6 +745,39 @@ public function links()
+            throw new \InvalidArgumentException('The current node list is empty.');
+        }
+
+        $node = $this->getNode(0);
```
We should make sure `$node` is an instance of `\DOMElement`, just like we do in the `link()` method.

jakzal commented on src/Symfony/Component/DomCrawler/Crawler.php:
```diff
@@ -731,6 +745,39 @@ public function links()
+    {
+        $images = array();
+        foreach ($this as $node) {
+            $images[] = new Image($node, $this->baseHref);
```
We should make sure `$node` is an instance of `\DOMElement`, just like we do in the `links()` method.

jakzal commented:
Final remark: why do we need the `UriElementInterface`?

fabpot commented:
Indeed, I would remove the interface.

stloyd commented on src/Symfony/Component/DomCrawler/AbstractUriElement.php:
```diff
@@ -0,0 +1,214 @@
+    protected $node;
+
+    /**
+     * @var string The method to use for the link
```
`The method to use for the URI element`

stloyd commented on src/Symfony/Component/DomCrawler/AbstractUriElement.php:
```diff
@@ -0,0 +1,214 @@
+    protected $method;
+
+    /**
+     * @var string The URI of the page where the link is embedded (or the base href)
```
`The URI of the page where the URI element is embedded (or the base href)`

valeriangalliat commented:
@jakzal @fabpot Not sure why I added the interface, I just removed it.

Also updated the PR according to comments. Just the `Constructor.` comment to be removed, but since it's present in files not affected by this PR, I aggree with @stof [that this should be done in another PR](https://github.com/symfony/symfony/pull/17585#discussion_r51733371).

jakzal commented:
Thanks @valeriangalliat for working on this feature, this is much appreciated.
